src/shared/vscode/commands2.ts

```diff
@@ -365,44 +370,215 @@ interface CommandInfo<T extends Callback> {
      * (default: 5 minutes)
      */
     readonly telemetryThrottleMs?: number | false
+
+    /**
+     * The indexes of the args that will be used to determine
+     * the uniqueness of the metric.
+     *
+     * Eg: If given `[0]` then
+     * the first argument in the `execute()` call will differentiate
+     * the metric from other metrics with the same `id`.
+     *
+     * The index starts at 0.
+     */
+    readonly uniqueArgIndexes?: UniqueArgIndex[]
+}
+
+/** 
+ * Used by {@link CommandInfo}, will indicate that the arg should be used
+ * as a telemetry attribute.
+ */
+export class IndexedAttribute {
+    /**
+     * 
+     * @param index The index of the arg this applies to
+     * @param attribute The name of the attribute that will be part of the `vscode_executeCommand` event
+     */
+    constructor(readonly index: number, readonly attribute: 'source') {}
+
+    /**
+     * Validates the given arg matches the expected type for the attribute.
+     * Throws if the type is unexpected.
+     */
+    validate(arg: any) {
+        switch (this.attribute) {
+            case 'source':
+                if (typeof arg !== 'string') {
+                    throw new ToolkitError(
```

Comment:
Why validate? Is this for defensiveness? If so, we can add test coverage for specific commands.

But I'm not sure that validation belongs here. In theory, any arg can be part of a key if the command definition wants it to be.

Reply:
`IndexAttribute` has string literals for `IndexAttribute.attribute`. The intent is for any arg to **not** be any value it wants to be, instead it must match the type decided in `IndexAttribute.validate()`.

Eg: `source` is meant to be the "source" attribute in `vscode_executeCommand` which is always a string

> Why validate? Is this for defensiveness? If so, we can add test coverage for specific commands.

Yeah for defensiveness. This is intended as a backup just in case the type is wrong and the implementation is not tested, otherwise a mistake can send invalid data

Reply:
Tests are the way to go here. Or the type system. Keep it simple.

Reply:
Deleted